Iteration3/master/app.py

```diff
@@ -0,0 +1,155 @@
+            )
+        message: str = request.json.get("message")
+
+        done = asyncio.Event()
+        done.write_concern = write_concern
+
+        MESSAGES.append(message)
+
+        if write_concern == 1:
+            for ip_address in slaves_ip_addresses():
+                send_replication_request(ip_address, message, MESSAGE_ID.value)
+            MESSAGE_ID.increment()
+            return jsonify({"message": "successful"})
+
+        common_tasks: list[tuple[asyncio.Task, str]] = []
+        for ip_address in slaves_ip_addresses():
+            task = asyncio.create_task(
+                replicate_on_slaves(
+                    ip_address, message, done, MESSAGE_ID.value
+                )
+            )
+            logger.info(
+                "Task for replication on slave:%s with ID:%d has been created",
+                ip_address,
+                MESSAGE_ID.value,
+            )
+            common_tasks.append((task, ip_address))
+        MESSAGE_ID.increment()
+
+        await done.wait()
+
+        # For tasks that were not finished we create threads in which
+        # we will make new requests to slaves for replication to ensure that
+        # slave has received our message.
+        if done.write_concern > 1:
+            for task, ip_address in common_tasks:
+                if not task.done():
+                    asyncio.create_task(do_retry_request(ip_address, message))
+    return ",".join(MESSAGES)
+
+
+async def do_retry_request(ip_address, message):
+    retry_cnt = 0
+    while (retry_cnt < MY_RETRY) and (
+        send_target(ip_address, message, MESSAGE_ID.value) != 200
+    ):
+        retry_cnt += 1
+        logger.error(
+            "Retried post to %s | try %d from %d",
+            ip_address,
+            retry_cnt,
+            MY_RETRY,
+        )
+
+
+async def replicate_on_slaves(
+    ip_address: str, message: str, done: asyncio.Event, message_id: int
+):
+    async with httpx.AsyncClient() as client:
+        try:
+            response = await client.post(
+                ip_address,
+                json={"message": message, "message_id": message_id},
+            )
+            logger.info(
+                "Response from the slave:%s with ID:%d has been received",
+                ip_address,
+                message_id,
+            )
+
+            # Basically, the next thing is callback of a coroutine.
+            if response.status_code == 200:
+                done.write_concern -= 1
+            if done.write_concern <= 1:
+                done.set()
+
+            return {
+                "ip_address": f"{ip_address}",
+                "status": response.status_code,
+            }
+        except (httpx.ConnectError, httpx.ReadTimeout) as exception:
+            done.set()
+            logger.error(
+                "Error when making request to the slave:%s with ID:%d %r, "
+                "sent for retry",
```

Comment:
Actually we are not sending it for retry immediately, so maybe should change logger message to something like "scheduled for the retry"

Reply:
Good point

Reply:
resolved in c78054bc0a202ffb1e889cd723865e480abdcffc